mautil/mautil.go

```diff
@@ -13,45 +13,18 @@ import (
 // loopback, or unspecified IP multiaddrs removed. If no multiaddrs are
 // removed, then returns the original slice.
 func FilterPrivateIPs(maddrs []multiaddr.Multiaddr) []multiaddr.Multiaddr {
-	var pvt []int
-	for i, maddr := range maddrs {
-		if maddr == nil {
-			continue
-		}
-		ip, err := manet.ToIP(maddr)
-		if err != nil {
-			continue
-		}
-		if ip.IsPrivate() || ip.IsLoopback() || ip.IsUnspecified() {
-			pvt = append(pvt, i)
-		}
-	}
-
-	// If no private addrs, return original slice.
-	if len(pvt) == 0 {
-		return maddrs
-	}
-
-	// If no non-private addrs, return nothing.
-	notPrivLen := len(maddrs) - len(pvt)
-	if notPrivLen == 0 {
+	filtered := multiaddr.FilterAddrs(maddrs, notPrivateAddr)
```

Comment:
Similar to the approach I have implemented [here](https://github.com/ipni/caskadht/pull/21), I think something like this should suffice:

```suggestion
	filtered := multiaddr.FilterAddrs(maddrs, func(target multiaddr.Multiaddr) bool {
		c, _ := multiaddr.SplitFirst(target)
		if c == nil {
			return false
		}
		switch c.Protocol().Code {
		case multiaddr.P_IP4, multiaddr.P_IP6, multiaddr.P_IP6ZONE, multiaddr.P_IPCIDR:
			return manet.IsPublicAddr(target)
		case multiaddr.P_DNS, multiaddr.P_DNS4, multiaddr.P_DNS6, multiaddr.P_DNSADDR:
			return c.Value() != "localhost"
		default:
			return true
		}
	})
```